Wrong grid track sizes in dev tools when using custom properties

AnujPanta1 opened the pull request:
#### 989c9f9cd5b1f0c9606820e219ee51da32a34c6b
<pre>
Wrong grid track sizes in dev tools when using custom properties
<a href="https://bugs.webkit.org/show_bug.cgi?id=284490">https://bugs.webkit.org/show_bug.cgi?id=284490</a>
<a href="https://rdar.apple.com/problem/141709306">rdar://problem/141709306</a>

Reviewed by Devin Rousso.

We now detect CSS variable references and resolve them while preserving authored values.
By using Style::Extractor::propertyValueInStyle() with a nullptr renderer, we avoid pixel
conversion that occurs for rendered grids.
This preserves track sizing values like 1fr, minmax(), repeat(), fit-content(), and
percentages while still resolving variable references.

* Source/WebCore/inspector/InspectorOverlay.cpp:
(WebCore::authoredGridTrackSizes):

Canonical link: <a href="https://commits.webkit.org/302811@main">https://commits.webkit.org/302811@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7d790ce57f70bb8203ecd4e0a6f1e2b20f7da6cf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/129290 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/159/builds/1546 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/138/builds/40127 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/136666 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/80680 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/131161 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/155/builds/1476 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/156/builds/1423 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/136666 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/80680 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/132237 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/155/builds/1476 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/138/builds/40127 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/136666 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/155/builds/1476 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/138/builds/40127 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/79944 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/155/builds/1476 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/138/builds/40127 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/139139 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/153/builds/1335 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/156/builds/1423 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/139139 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/160/builds/1379 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/138/builds/40127 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/139139 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27405 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/138/builds/40127 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/53957 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/158/builds/1407 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/64765 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/1225 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/1260 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/1329 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebCore/inspector/InspectorOverlay.cpp:
```diff
@@ -1434,10 +1435,24 @@ static Vector<String> authoredGridTrackSizes(Node* node, Style::GridTrackSizingD
+        auto variableText = cssValue->cssText(CSS::defaultSerializationContext());
+        auto computedStyle = CSSComputedStyleDeclaration::create(*element, CSSComputedStyleDeclaration::AllowVisited::No);
+        auto computedValue = computedStyle->getPropertyValue(directionCSSPropertyID);
+        auto variableName = variableText.substring(4, variableText.length() - 5);
```
is the `4` because of `var(`?  if so, i think this would be nicer as a separate variable so it's more clear
```cpp
static constexpr auto varPrefix = "var("_s;
static constexpr auto varSuffix = ")"_s;
auto variableName = variableText.substring(varPrefix.length(), variableText.length() - varPrefix.length() - varSuffix.length());
```

also, does `hasVariableReferences` guarantee that the top-level value is a `var()`?  or should we check `variableText.startsWith("var("_s)` and bail if not?

dcrousso commented on Source/WebCore/inspector/InspectorOverlay.cpp:
```diff
@@ -1434,10 +1435,24 @@ static Vector<String> authoredGridTrackSizes(Node* node, Style::GridTrackSizingD
+        else
+            trackSizes.append(variableText);
+
+        return trackSizes;
```
rather than creating an empty vector and appending to it, could we do something like this?
```cpp
if (!computedValue.isEmpty() && computedValue != variableText)
    return { makeString(variableName, ": "_s, computedValue) };
return { variableText };
```

dcrousso commented on Source/WebCore/inspector/InspectorOverlay.cpp:
```diff
@@ -1434,10 +1435,24 @@ static Vector<String> authoredGridTrackSizes(Node* node, Style::GridTrackSizingD
     auto* cssValueList = dynamicDowncast<CSSValueList>(cssValue.get());
     if (!cssValueList)
         return { };
-    
```
oops?

dcrousso reviewed (commented):
can you please attach some before/after screenshots so we can see the effect?

dcrousso commented on Source/WebCore/inspector/InspectorOverlay.cpp:
```diff
@@ -1434,10 +1435,17 @@ static Vector<String> authoredGridTrackSizes(Node* node, Style::GridTrackSizingD
 
+    if (cssValue && cssValue->hasVariableReferences()) {
+        Style::Extractor extractor(element);
+        auto computedValue = extractor.propertyValue(directionCSSPropertyID);
```
nice! this is much cleaner 😃

thinking about this a bit more, i think it might actually be nice to show that the value is originally `auto` so that developers have some visual clue as to what things may change if the container adjusts

perhaps something like `auto - 100px` or _`(100px)`_ (either italicized and/or in a lightgrey color to indicate it's not a "normal" value)?

since this PR is already an improvement as-is i'm ok doing this as a followup if you'd rather keep things cleaner (and if so please just file a bug with a FIXME comment here)

AnujPanta1 replied:
I'll do a fix me for simplicity sake thank you

patrickangle commented on Source/WebCore/inspector/InspectorOverlay.cpp:
Does this actually get the computed value, or does it just resolve the variable? The overlay is meant to show you the real values, but those aren't necessarily the computed value. For example `1fr` could end up computed to be `100px`, but we'd want to show `1fr` I think to match what we'd show a track that whose size was explicitly defined as `1fr`?

dcrousso replied:
ooooo good point!

grid has too many special units ☹️

AnujPanta1 replied:
I'm going to do more testing on this before resolving

dcrousso commented on Source/WebCore/inspector/InspectorOverlay.cpp:
```diff
@@ -1434,10 +1435,17 @@ static Vector<String> authoredGridTrackSizes(Node* node, Style::GridTrackSizingD
         }
     }
 
+    if (cssValue && cssValue->hasVariableReferences()) {
```
i wonder if maybe we should always do something this in case the value has a `calc` since something like `calc(1em * 42px)` is somewhat inscrutable

or are `calc` already resolved through some other mechanism by this point?

AnujPanta1 replied:
From the test page I had made I believe I saw calc resolved at this point

dcrousso reviewed (approved):
nice work!

AnujPanta1 committed:
Wrong grid track sizes in dev tools when using custom properties
https://bugs.webkit.org/show_bug.cgi?id=284490
rdar://problem/141709306

Reviewed by Devin Rousso.

We now detect CSS variable references and resolve them while preserving authored values.
By using Style::Extractor::propertyValueInStyle() with a nullptr renderer, we avoid pixel
conversion that occurs for rendered grids.
This preserves track sizing values like 1fr, minmax(), repeat(), fit-content(), and
percentages while still resolving variable references.

* Source/WebCore/inspector/InspectorOverlay.cpp:
(WebCore::authoredGridTrackSizes):

Canonical link: https://commits.webkit.org/302811@main
